### PR TITLE
feat(chronicleexporter): Support dynamic namespace and ingestion labels 

### PR DESCRIPTION
**Pull request by justinianvoss22**
### Proposed Change
- Update chronicle exporter to be able to take namespace and ingestion labels from attributes if present.
- Make logType not required

### Testing
- Unit testing
- Testing agent on local BP

##### Checklist
- [ x ] Changes are tested
- [ x ] CI has passed

### CODE REVIEW

**Comment by jsirianni** [comment]
I updated the PR name, it will help w/ release notes. Feel free to tweak it if its incorrect.

**Comment by kuiperda** [comment]
Took a quick high level pass and it makes sense to me, will let Dakota review though.

File: exporter/chronicleexporter/marshal.go [hunk 1]
```diff
@@ -50,7 +51,6 @@ type logMarshaler interface {
 	MarshalRawLogs(ctx context.Context, ld plog.Logs) ([]*api.BatchCreateLogsRequest, error)
 	MarshalRawLogsForHTTP(ctx context.Context, ld plog.Logs) (map[string]*api.ImportLogsRequest, error)
 }
-
 type protoMarshaler struct {
```
Comment (dpaasman00) [inline exporter/chronicleexporter/marshal.go:54]: Since we're using dynamic labels, the `labels` object on this struct is no longer used and should be removed.

File: exporter/chronicleexporter/marshal.go [hunk 2]
```diff
@@ -73,64 +72,89 @@ func newProtoMarshaler(cfg Config, teleSettings component.TelemetrySettings, lab
...
+					namespaceMap[logType] = namespace
+				}
+				if len(ingestionLabels) != 0 {
+					if ingestionLabelsMap[logType] == nil {
+						ingestionLabelsMap[logType] = make([]*api.Label, 0)
+					}
+					for _, v := range ingestionLabels {
+						ingestionLabelsMap[logType] = append(ingestionLabelsMap[logType], v)
+					}
+				}
```
Comment (dpaasman00) [inline exporter/chronicleexporter/marshal.go:122]: I think some more thought needs to go into how these labels get added. Currently we append all labels present on a log record to the overarching map, regardless of if they're already present. This causes repeat entries if there are multiple log records with the same labels. You can add this test case to the test `TestProtoMarshaler_MarshalRawLogs` in `marshal_test.go` test to see this behavior:
```go
{
			name: "Multiple log records with attributes",
			cfg: Config{
				CustomerID:      uuid.New().String(),
				LogType:         "WINEVTLOG",
				Namespace:       "test",
				RawLogField:     "body",
				OverrideLogType: false,
			},
			labels: []*api.Label{},
			logRecords: func() plog.Logs {
				logs := plog.NewLogs()
				record1 := logs.ResourceLogs().AppendEmpty().ScopeLogs().AppendEmpty().LogRecords().AppendEmpty()
				record1.Body().SetStr("First log message")
				record1.Attributes().FromRaw(map[string]any{`chronicle_ingestion_label["key1"]`: "value1", `chronicle_ingestion_label["key2"]`: "value2"})
				record2 := logs.ResourceLogs().AppendEmpty().ScopeLogs().AppendEmpty().LogRecords().AppendEmpty()
				record2.Body().SetStr("Second log message")
				record2.Attributes().FromRaw(map[string]any{`chronicle_ingestion_label["key1"]`: "value1", `chronicle_ingestion_label["key2"]`: "value2"})
				return logs
			},
			expectations: func(t *testing.T, requests []*api.BatchCreateLogsRequest) {
				require.Len(t, requests, 1, "Expected a single batch request")
				batch := requests[0].Batch
				// verify batch source labels
				require.Len(t, batch.Source.Labels, 2)
				require.Len(t, batch.Entries, 2, "Expected two log entries in the batch")
				// Verifying the first log entry data
				require.Equal(t, "First log message", string(batch.Entries[0].Data))
				// Verifying the second log entry data
				require.Equal(t, "Second log message", string(batch.Entries[1].Data))
			},
		},
```

File: exporter/chronicleexporter/marshal_test.go
Comment (dpaasman00) [inline exporter/chronicleexporter/marshal_test.go:1]: I think there are some more test cases that can be added to thoroughly test this new functionality and potential edge cases, such as the multiple log records example I gave in an earlier comment.

File: exporter/chronicleexporter/marshal.go
Comment (dpaasman00) [inline exporter/chronicleexporter/marshal.go:1]: This is a general, very nit comment. Unless you feel strongly about it, there are several deleted new lines in this PR that I think should be restored to help with readability. Just take a look through the diff here to find them.

File: exporter/chronicleexporter/marshal.go [hunk 3]
```diff
@@ -61,7 +61,6 @@ type protoMarshaler struct {
 }
 
 func newProtoMarshaler(cfg Config, teleSettings component.TelemetrySettings, labels []*api.Label, customerID []byte) (*protoMarshaler, error) {
-
 	return &protoMarshaler{
 		startTime:    time.Now(),
```
Comment (dpaasman00) [inline exporter/chronicleexporter/marshal.go:65]: Contrary to the deleted newlines comment I left, this is a valid deleted new line and should stay deleted.

**Review by dpaasman00 — CHANGES_REQUESTED** [review]
First pass and generally looks good. Somethings to address however.

File: exporter/chronicleexporter/marshal.go [hunk 4]
```diff
@@ -177,6 +215,35 @@ func (m *protoMarshaler) getLogType(ctx context.Context, logRecord plog.LogRecor
...
+	}
+	ingestionLabels, err := m.getRawNestedFields(chronicleIngestionLabelsPrefix, logRecord)
+	if err != nil {
+		return configLabels, fmt.Errorf("get chronicle ingestion labels: %w", err)
+	}
+	if len(ingestionLabels) != 0 {
+		return ingestionLabels, nil
+	}
+	return configLabels, nil
+}
```
Comment (dpaasman00) [inline exporter/chronicleexporter/marshal.go:245]: Again, the return value isn't used when there's an error, so return an empty slice and simplify this code a bit
```suggestion
func (m *protoMarshaler) getIngestionLabels(logRecord plog.LogRecord) ([]*api.Label, error) {
	// check for labels in attributes["chronicle_ingestion_labels"]
	ingestionLabels, err := m.getRawNestedFields(chronicleIngestionLabelsPrefix, logRecord)
	if err != nil {
		return []*api.Label{}, fmt.Errorf("get chronicle ingestion labels: %w", err)
	}
	if len(ingestionLabels) != 0 {
		return ingestionLabels, nil
	}

	// use labels defined in config if needed
	configLabels := make([]*api.Label, 0)
	for key, value := range m.cfg.IngestionLabels {
		configLabels = append(configLabels, &api.Label{
			Key:   key,
			Value: value,
		})
	}

	return configLabels, nil
}
```

File: exporter/chronicleexporter/marshal.go [hunk 5]
```diff
@@ -207,10 +274,32 @@ func (m *protoMarshaler) getRawField(ctx context.Context, field string, logRecor
...
+			nestedFields = append(nestedFields, &api.Label{
+				Key:   key,
+				Value: value.AsString(),
+			})
+		}
+		return true
+	})
+	if len(nestedFields) == 0 {
+		return nil, nil
+	}
```
Comment (dpaasman00) [inline exporter/chronicleexporter/marshal.go:290]: No need to return nil instead of an empty `[]*api.Label`

```suggestion
```

File: exporter/chronicleexporter/marshal.go
Comment (dpaasman00) [inline exporter/chronicleexporter/marshal.go:1]: I know your PR doesn't touch the `getLogType()` function, but in the spirit of adding 2 similar funcs, can you update `getLogType()` to return an empty string in the 2 cases it returns an error, because the returned string does not get used when there's an error

File: exporter/chronicleexporter/README.md [hunk 6]
```diff
@@ -48,8 +48,9 @@ currently supported log types are:
 - windows_event.system
 - sql_server
 
-
 If the `attributes["chronicle_log_type"]` field is present in the log, we will use its value in the payload instead of the automatic detection or the `log_type` in the config.
+If the `attributes["chronicle_namespace"]` field is present in the log, we will use its value in the payload.
+If there are nested fields in `attributes["chronicle_ingestion_label"]`, we will use the values in the payload.
```
Comment (dpaasman00) [inline exporter/chronicleexporter/README.md:53]: These should be broken out into a separate section from the "Log Type" section they're in currently, something like "Namespace & Ingestion Labels". Also state that if the attribute fields are present, they are used regardless of the value in the config.

File: exporter/chronicleexporter/marshal.go [hunk 7]
```diff
@@ -113,14 +114,54 @@ func (m *protoMarshaler) extractRawLogs(ctx context.Context, ld plog.Logs) (map[
 					Data:           []byte(rawLog),
 				}
 				entries[logType] = append(entries[logType], entry)
+				// each logType maps to exactly 1 namespace value
+				if _, ok := namespaceMap[logType]; !ok {
+					namespaceMap[logType] = namespace
+				}
```
Comment (dpaasman00) [inline exporter/chronicleexporter/marshal.go:120]: This should also check if the namespace is empty before adding
```suggestion
                         if namespace != "" {
				// each logType maps to exactly 1 namespace value
				if _, ok := namespaceMap[logType]; !ok {
					namespaceMap[logType] = namespace
				}
			}
```

**Review by dpaasman00 — CHANGES_REQUESTED** [review]
Couple more thoughts, otherwise this is looking good. Going to test next.

**Review by dpaasman00 — APPROVED** [review]
This has been working well for me.
